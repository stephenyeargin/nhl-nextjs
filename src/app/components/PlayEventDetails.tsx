'use client';

import React from 'react';
import Link from 'next/link';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheckCircle, faPlayCircle } from '@fortawesome/free-solid-svg-icons';
import Headshot from './Headshot';
import TeamLogo from './TeamLogo';
import {
  GAME_EVENTS,
  MISS_TYPES,
  NHL_BRIGHTCOVE_ACCOUNT,
  PENALTY_DESCRIPTIONS,
  PENALTY_TYPES,
  ZONE_DESCRIPTIONS,
} from '@/app/utils/constants';
import { formatPeriodLabel } from '@/app/utils/formatters';

type RosterPlayer = {
  playerId?: number;
  sweaterNumber?: number;
  headshot?: string;
  firstName?: { default?: string };
  lastName?: { default?: string };
};

type TeamColors = {
  teamColor?: string;
  abbreviation?: string;
};

export interface PlayEventDetailsProps {
  play: any;
  game: any;
  rosterSpots: RosterPlayer[];
  lookupTeamDataByTeamId: (_teamId: number) => TeamColors;
  onOpenHighlight?: (_payload: { url: string; label: string }) => void;
}

const PlayEventDetails: React.FC<PlayEventDetailsProps> = ({
  play,
  game,
  rosterSpots,
  lookupTeamDataByTeamId,
  onOpenHighlight,
}) => {
  const e = play.details || {};

  const lookupPlayerData = (playerId: number): RosterPlayer => {
    const defPlayer = { firstName: { default: 'Unnamed' }, lastName: { default: 'Player' } };
    const found = rosterSpots.find((pl: any) => pl.playerId === playerId);

    return (found || (defPlayer as RosterPlayer)) as RosterPlayer;
  };

  const renderPlayer = (playerId: number) => {
    const player = lookupPlayerData(playerId);
    if (!player.playerId) {
      return (
        <>
          <span className="font-bold">
            {player.firstName?.default} {player.lastName?.default}
          </span>
        </>
      );
    }

    return (
      <>
        <span className="p-1 border rounded text-xs mx-1">#{player.sweaterNumber}</span>{' '}
        <Link href={`/player/${player.playerId}`} className="font-bold">
          {player.firstName?.default} {player.lastName?.default}
        </Link>
      </>
    );
  };

  const eventTeamData: any = lookupTeamDataByTeamId(e.eventOwnerTeamId);

  switch (play.typeDescKey) {
    case 'period-start':
      return (
        <div>
          Start of{' '}
          {formatPeriodLabel(
            { ...game.periodDescriptor, number: play.periodDescriptor.number },
            true
          )}
        </div>
      );
    case 'faceoff':
      return (
        <div>
          {renderPlayer(e.winningPlayerId)} won a faceoff against {renderPlayer(e.losingPlayerId)}{' '}
          {e.zoneCode ? `in the ${(ZONE_DESCRIPTIONS as Record<string, string>)[e.zoneCode]}` : ''}
        </div>
      );
    case 'blocked-shot':
      return (
        <div>
          {renderPlayer(e.shootingPlayerId)}&apos;s shot was blocked by{' '}
          {e.reason === 'teammate-blocked' ? 'teammate' : ''} {renderPlayer(e.blockingPlayerId)}
        </div>
      );
    case 'shot-on-goal':
      return (
        <div>
          {renderPlayer(e.shootingPlayerId)} took a shot on {renderPlayer(e.goalieInNetId)}
        </div>
      );
    case 'stoppage':
      return (
        <div>
          {(GAME_EVENTS as Record<string, string>)[e.reason]}
          {e.secondaryReason && e.secondaryReason !== e.reason
            ? `, ${(GAME_EVENTS as Record<string, string>)[e.secondaryReason]}`
            : ''}
        </div>
      );
    case 'hit':
      return (
        <div>
          {renderPlayer(e.hittingPlayerId)} hit {renderPlayer(e.hitteePlayerId)}{' '}
          {e.zoneCode ? `in the ${(ZONE_DESCRIPTIONS as Record<string, string>)[e.zoneCode]}` : ''}
        </div>
      );
    case 'giveaway':
      return (
        <div>
          {renderPlayer(e.playerId)} gave the puck away{' '}
          {e.zoneCode ? `in the ${(ZONE_DESCRIPTIONS as Record<string, string>)[e.zoneCode]}` : ''}
        </div>
      );
    case 'takeaway':
      return (
        <div>
          {renderPlayer(e.playerId)} took the puck away{' '}
          {e.zoneCode ? `in the ${(ZONE_DESCRIPTIONS as Record<string, string>)[e.zoneCode]}` : ''}
        </div>
      );
    case 'missed-shot':
      return (
        <div>
          {renderPlayer(e.shootingPlayerId)} missed a {e.shotType} shot{' '}
          {e.reason ? `(${(MISS_TYPES as Record<string, string>)[e.reason] || e.reason})` : ''}
        </div>
      );
    case 'goal':
      return (
        <div
          className="p-2 gap-5 rounded-lg flex flex-wrap justify-center md:justify-between items-center bg-red-900/80 text-white"
          style={{ borderLeft: `solid 25px ${eventTeamData.teamColor}` }}
        >
          <div className="flex gap-5 items-center justify-start">
            <Headshot
              playerId={e.scoringPlayerId}
              src={lookupPlayerData(e.scoringPlayerId).headshot}
              alt="Player Headshot"
              className="h-16 w-16 hidden lg:block"
              team={eventTeamData.abbreviation}
            />
            <div>
              <div className="font-medium text-lg mb-3">
                {renderPlayer(e.scoringPlayerId)}{' '}
                {e.scoringPlayerTotal ? `(${e.scoringPlayerTotal})` : ''} scored{' '}
                {e.shotType
                  ? `(${(GAME_EVENTS as Record<string, string>)[e.shotType]?.toLowerCase()})`
                  : ''}
              </div>
              {(e.assist1PlayerId || e.assist2PlayerId) && <span>Assisted By: </span>}
              {e.assist1PlayerId && (
                <span>
                  {renderPlayer(e.assist1PlayerId)} ({e.assist1PlayerTotal})
                </span>
              )}
              {e.assist2PlayerId && (
                <span>
                  {' '}
                  and {renderPlayer(e.assist2PlayerId)} ({e.assist2PlayerTotal})
                </span>
              )}
              {!e.assist1PlayerId && !e.assist2PlayerId && <span>Unassisted</span>}
            </div>
          </div>
          <div className="w-10">
            {play.periodDescriptor?.periodType === 'SO' ? (
              <div className="text-2xl text-yellow-500">
                <FontAwesomeIcon icon={faCheckCircle} />
              </div>
            ) : (
              <div className="text-2xl font-bold">
                {e.awayScore}-{e.homeScore}
              </div>
            )}
          </div>
          {play.details?.highlightClip && (
            <div className="text-center text-white">
              <button
                onClick={() => {
                  if (!onOpenHighlight) {
                    return;
                  }
                  const player = lookupPlayerData(e.scoringPlayerId);
                  const url = `https://players.brightcove.net/${NHL_BRIGHTCOVE_ACCOUNT}/default_default/index.html?videoId=${play.details.highlightClip}`;
                  const label = `${eventTeamData.abbreviation} | ${play.timeInPeriod} ${formatPeriodLabel(play.periodDescriptor)} | ${player.firstName?.default} ${player.lastName?.default}`;
                  onOpenHighlight({ url, label });
                }}
              >
                <FontAwesomeIcon icon={faPlayCircle} size="2x" className="align-middle mx-auto" />
              </button>
            </div>
          )}
        </div>
      );
    case 'penalty':
      return (
        <div
          className="p-2 gap-5 rounded-lg flex flex-wrap justify-center md:justify-start items-center bg-slate-500/80 text-white"
          style={{ borderLeft: `solid 25px ${eventTeamData.teamColor}` }}
        >
          {e.committedByPlayerId ? (
            <Headshot
              playerId={e.committedByPlayerId}
              src={lookupPlayerData(e.committedByPlayerId).headshot}
              alt="Player Headshot"
              className="h-16 w-16 hidden lg:block"
              team={eventTeamData.abbreviation}
            />
          ) : (
            <div className="bg-slate-100 rounded-full h-16 w-16 hidden lg:block">
              <TeamLogo team={eventTeamData.abbreviation} className="h-16 w-16" />
            </div>
          )}
          <div>
            {e.committedByPlayerId ? (
              <div className="font-medium text-lg mb-3">{renderPlayer(e.committedByPlayerId)}</div>
            ) : (
              <div className="font-medium text-lg mb-3">Team Penalty</div>
            )}
            {e.servedByPlayerId && (
              <div className="text-xs my-2">Served by: {renderPlayer(e.servedByPlayerId)}</div>
            )}
            {e.drawnByPlayerId && (
              <div className="text-xs my-2">Drawn By: {renderPlayer(e.drawnByPlayerId)}</div>
            )}
          </div>
          <div>
            <div className="text-xs font-light">Duration</div>
            <div>{e.duration} minutes</div>
          </div>
          <div>
            <div className="text-xs font-light">
              {(PENALTY_TYPES as Record<string, string>)[e.typeCode]}
            </div>
            <div>{(PENALTY_DESCRIPTIONS as Record<string, string>)[e.descKey]}</div>
          </div>
        </div>
      );
    case 'period-end':
      return (
        <div>
          End of{' '}
          {formatPeriodLabel(
            { ...game.periodDescriptor, number: play.periodDescriptor.number },
            true
          )}
        </div>
      );
    case 'shootout-complete':
      return <div>Shootout complete</div>;
    case 'delayed-penalty':
      return <div>Delayed penalty</div>;
    case 'game-end':
      return <div>End of game</div>;
    default:
      return null;
  }
};

export default PlayEventDetails;
