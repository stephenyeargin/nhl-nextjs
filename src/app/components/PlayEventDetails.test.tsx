import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import PlayEventDetails from './PlayEventDetails';

jest.mock('./Headshot', () => {
  const Mock = ({ playerId, className }: any) => (
    <div data-testid={`headshot-${playerId}`} className={className} />
  );
  (Mock as any).displayName = 'HeadshotMock';

  return Mock;
});

jest.mock('./TeamLogo', () => {
  const Mock = (props: any) => <div data-testid="team-logo" {...props} />;
  (Mock as any).displayName = 'TeamLogoMock';

  return Mock;
});

const game = { periodDescriptor: { number: 1 } };
const rosterSpots = [
  {
    playerId: 9,
    sweaterNumber: 9,
    headshot: '/9.jpg',
    firstName: { default: 'Filip' },
    lastName: { default: 'Forsberg' },
  },
  {
    playerId: 92,
    sweaterNumber: 92,
    headshot: '/92.jpg',
    firstName: { default: 'Ryan' },
    lastName: { default: 'Johansen' },
  },
];

const lookupTeamDataByTeamId = (teamId: number) =>
  teamId === 1
    ? { teamColor: '#111', abbreviation: 'NSH' }
    : { teamColor: '#222', abbreviation: 'CBJ' };

describe('PlayEventDetails', () => {
  it('renders goal with highlight button and triggers callback', () => {
    const onOpenHighlight = jest.fn();
    const play = {
      typeDescKey: 'goal',
      details: {
        scoringPlayerId: 9,
        assist1PlayerId: 92,
        assist1PlayerTotal: 1,
        awayScore: 1,
        homeScore: 0,
        eventOwnerTeamId: 1,
        highlightClip: '12345',
      },
      timeInPeriod: '10:00',
      periodDescriptor: { number: 1 },
    };
    const { asFragment } = render(
      <PlayEventDetails
        play={play}
        game={game}
        rosterSpots={rosterSpots as any}
        lookupTeamDataByTeamId={lookupTeamDataByTeamId as any}
        onOpenHighlight={onOpenHighlight}
      />
    );
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(onOpenHighlight).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders penalty summary', () => {
    const play = {
      typeDescKey: 'penalty',
      details: {
        committedByPlayerId: 92,
        duration: 2,
        typeCode: 'minor',
        descKey: 'tripping',
        eventOwnerTeamId: 1,
      },
      periodDescriptor: { number: 1 },
    };
    const { asFragment } = render(
      <PlayEventDetails
        play={play}
        game={game}
        rosterSpots={rosterSpots as any}
        lookupTeamDataByTeamId={lookupTeamDataByTeamId as any}
      />
    );
    expect(screen.getByText(/minutes/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders faceoff details', () => {
    const play = {
      typeDescKey: 'faceoff',
      details: {
        winningPlayerId: 92,
        losingPlayerId: 9,
        zoneCode: 'O',
        eventOwnerTeamId: 2,
      },
      periodDescriptor: { number: 1 },
    };
    const { asFragment } = render(
      <PlayEventDetails
        play={play}
        game={game}
        rosterSpots={rosterSpots as any}
        lookupTeamDataByTeamId={lookupTeamDataByTeamId as any}
      />
    );
    expect(screen.getByText(/won a faceoff/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders shootout goal without player total parentheses', () => {
    const play = {
      typeDescKey: 'goal',
      details: {
        scoringPlayerId: 9,
        awayScore: 0,
        homeScore: 1,
        eventOwnerTeamId: 1,
      },
      timeInPeriod: 'SO',
      periodDescriptor: { number: 5, periodType: 'SO' },
    };
    const { container, asFragment } = render(
      <PlayEventDetails
        play={play}
        game={game}
        rosterSpots={rosterSpots as any}
        lookupTeamDataByTeamId={lookupTeamDataByTeamId as any}
      />
    );
    // Name should render
    expect(screen.getByText('Filip Forsberg')).toBeInTheDocument();
    // Should not render a goals total like "(1)" after the name for SO
    expect(container).not.toHaveTextContent(/\(\d+\)/);
    expect(asFragment()).toMatchSnapshot();
  });
});
