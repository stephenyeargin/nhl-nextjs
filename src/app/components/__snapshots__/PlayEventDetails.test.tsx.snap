// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayEventDetails renders faceoff details 1`] = `
<DocumentFragment>
  <div>
    <span
      class="p-1 border rounded text-xs mx-1"
    >
      #92
    </span>
     
    <a
      class="font-bold"
      href="/player/92"
    >
      Ryan Johansen
    </a>
     won a faceoff against 
    <span
      class="p-1 border rounded text-xs mx-1"
    >
      #9
    </span>
     
    <a
      class="font-bold"
      href="/player/9"
    >
      Filip Forsberg
    </a>
     in the offensive zone
  </div>
</DocumentFragment>
`;

exports[`PlayEventDetails renders goal with highlight button and triggers callback 1`] = `
<DocumentFragment>
  <div
    class="p-2 gap-5 rounded-lg flex flex-wrap justify-center md:justify-between items-center bg-red-900/80 text-white"
    style="border-left: 25px solid #111;"
  >
    <div
      class="flex gap-5 items-center justify-start"
    >
      <div
        class="h-16 w-16 hidden lg:block"
        data-testid="headshot-9"
      />
      <div>
        <div
          class="font-medium text-lg mb-3"
        >
          <span
            class="p-1 border rounded text-xs mx-1"
          >
            #9
          </span>
           
          <a
            class="font-bold"
            href="/player/9"
          >
            Filip Forsberg
          </a>
            scored 
        </div>
        <span>
          Assisted By: 
        </span>
        <span>
          <span
            class="p-1 border rounded text-xs mx-1"
          >
            #92
          </span>
           
          <a
            class="font-bold"
            href="/player/92"
          >
            Ryan Johansen
          </a>
           (1)
        </span>
      </div>
    </div>
    <div
      class="w-10"
    >
      <div
        class="text-2xl font-bold"
      >
        1-0
      </div>
    </div>
    <div
      class="text-center text-white"
    >
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-play fa-2x align-middle mx-auto"
          data-icon="circle-play"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256zM188.3 147.1c-7.6 4.2-12.3 12.3-12.3 20.9l0 176c0 8.7 4.7 16.7 12.3 20.9s16.8 4.1 24.3-.5l144-88c7.1-4.4 11.5-12.1 11.5-20.5s-4.4-16.1-11.5-20.5l-144-88c-7.4-4.5-16.7-4.7-24.3-.5z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlayEventDetails renders penalty summary 1`] = `
<DocumentFragment>
  <div
    class="p-2 gap-5 rounded-lg flex flex-wrap justify-center md:justify-start items-center bg-slate-500/80 text-white"
    style="border-left: 25px solid #111;"
  >
    <div
      class="h-16 w-16 hidden lg:block"
      data-testid="headshot-92"
    />
    <div>
      <div
        class="font-medium text-lg mb-3"
      >
        <span
          class="p-1 border rounded text-xs mx-1"
        >
          #92
        </span>
         
        <a
          class="font-bold"
          href="/player/92"
        >
          Ryan Johansen
        </a>
      </div>
    </div>
    <div>
      <div
        class="text-xs font-light"
      >
        Duration
      </div>
      <div>
        2 minutes
      </div>
    </div>
    <div>
      <div
        class="text-xs font-light"
      />
      <div>
        Tripping
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlayEventDetails renders shootout goal without player total parentheses 1`] = `
<DocumentFragment>
  <div
    class="p-2 gap-5 rounded-lg flex flex-wrap justify-center md:justify-between items-center bg-red-900/80 text-white"
    style="border-left: 25px solid #111;"
  >
    <div
      class="flex gap-5 items-center justify-start"
    >
      <div
        class="h-16 w-16 hidden lg:block"
        data-testid="headshot-9"
      />
      <div>
        <div
          class="font-medium text-lg mb-3"
        >
          <span
            class="p-1 border rounded text-xs mx-1"
          >
            #9
          </span>
           
          <a
            class="font-bold"
            href="/player/9"
          >
            Filip Forsberg
          </a>
            scored 
        </div>
        <span>
          Unassisted
        </span>
      </div>
    </div>
    <div
      class="w-10"
    >
      <div
        class="text-2xl text-yellow-500"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-check "
          data-icon="circle-check"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM369 209L241 337c-9.4 9.4-24.6 9.4-33.9 0l-64-64c-9.4-9.4-9.4-24.6 0-33.9s24.6-9.4 33.9 0l47 47L335 175c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
